src/exercises/assignBookmarksToExercises.js

```diff
@@ -36,16 +36,27 @@ function assignBookmarksWithLearningCycle(bookmarks, exerciseTypesList) {
 
       // Check if there are enough bookmarks for the selected exercise
       if (i + selectedExercise.requiredBookmarks <= bookmarks.length) {
-        let exercise = {
-          type: selectedExercise.type,
-          bookmarks: bookmarks.slice(i, i + selectedExercise.requiredBookmarks),
-        };
-        exerciseSequence.push(exercise);
+        let potentialBookmarks = bookmarks.slice(
+          i,
+          i + selectedExercise.requiredBookmarks,
+        );
+        let contexts = potentialBookmarks.map((bookmark) => bookmark.context);
+        let distinctContexts = new Set(contexts).size;
+        if (distinctContexts === contexts.length) {
+          let exercise = {
+            type: selectedExercise.type,
+            bookmarks: potentialBookmarks,
+          };
+          exerciseSequence.push(exercise);
```

Comment:
I have a hard time understanding this code. I've pushed a little commit with a clarification but it's still hard. @tfnribeiro  - can you have a look and see if you understand it? It might be just me :)

I suspect that extracting some helper methods might help? And maybe renaming some variables for clarity.

Reply:
I coded this with Merle, so I might be a bit biased. 

Essentially, we are checking for unique contexts within that slice of bookmarks whenever we pick an exercise with multiple bookmarks, and if the bookmarks aren't distinct it doesn't add the exercise. 

A suggestion for readability might be to call the code from line 39 -> 44, `hasDistinctContexts()` and then the check is based on that rather than that conditional.

Reply:
Oki, I've taken your suggestion, and tried out a few other renames to help along. 
One line that is still confusing for me is:

```python
    let possibleExercisesTypes =
      exercisesByType[learningCycleEnum[bookmarks[i].learning_cycle]];
```

`exercisesByType` - should it be named `exercisesByLearningCycle`? Both the var and the corresponding function (i.e., `getExerciseListByType`). 

Finally, in my mind it's a bit confusing when should we be talking about `exerciseType` and `exercise`. I've put `exerciseType` where I thought it was appropriate, but I think it might be missing elsewhere. And I think you might have used `exercise` to implicitly mean `exerciseType`. 

Can you do another pass over this function and make sure that my changes make sense? Thanks!

Reply:
I think there should be a differentiation. To me:

**Exercise Type**: Productive VS Receptive
**Exercise**: Match, Cloze, etc

But I can see that can create some confusion, so lets come up with a name we can all agree. I don't think Type is useful unless it is used to refer to Productive VS Receptive, so maybe we should just have:

**ExerciseListForCycle** (as opposed to _possibleExercisesTypes_): Productive or Receptive Type list
**ExerciseType** -> **Exercise**, like you suggest I don't think we loose much by removing the type essentially it should just be an instance of a single exercise.

Reply:
I agree that `exercisesByType` should be `exercisesByLearningCycle` as Mircea suggests and we can also change `possibleExerciseTypes` to `exerciseListForCycle`.

In general, I think the issue is that we are calling the different exercises (Match, Cloze etc.) ExerciseTypes everywhere else. That wasn't confusing before we introduced the learning cycle, but might be confusing now. I'd say something like

**ExerciseCycle**: productive vs receptive
**ExerciseType**: Match, Cloze..

Reply:
I'm on it :)